accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+        public bool TryGetInt16(out short value);
+        public bool TryGetInt32(out int value);
+        public bool TryGetInt64(out long value);
+        public bool TryGetSByte(out sbyte value);
+        public bool TryGetSingle(out float value);
+        public bool TryGetUInt16(out ushort value);
+        public bool TryGetUInt32(out uint value);
+        public bool TryGetUInt64(out ulong value);
+    }
+}
+```
+
+## Serializer interop
+Unlike the 5.0 `JsonNode` prototype and the existing `JsonDocument`, the `JsonNode` types are intended to be used by the serializer directly and do not contain `Parse()` or `Write()` methods themselves because:
+- The existing serializer methods can be used which support UTF8, string, Utf8JsonReader\Writer and Stream. Any future overloads will just work.
+- Serializer-specific functionality including custom converters and quoted numbers just work.
+
+The `JsonSerializerOptions` is passed to `JsonNode` and used to specify any run-time added custom converters, handle other options such as quoted numbers, and to respect the `PropertyNameCaseInsensitive` to determine whether the string-based lookup of property names on `JsonObject` are case-sensitive (for both direct use of `JsonObject` and also indirect use through `dynamic`). If we have `static JsonValue Parse()` methods, those would also need an override to specify the `JsonSerializerOptions`.
+
+Although the serializer can be used to directly (de)serialize `JsonNode` types, once a `JsonNode` instance is created, it is expected the `JsonNode.GetValue<T>()` method is used to change types instead of using the serializer:
+```cs
+    JsonNode node = ... // any value
+    string str;
+
+    // This works, but has no caching:
+    byte[] utf8 = JsonSerializer.SerializeToUtf8Bytes(node, options);
+    str = JsonSerializer.Deserialize<string>(utf8, options);
+
+    // For ease-of-use and performance, use GetValue().
+    // It may return the raw internal value or the last converted value instead of deserializing.
+    str = node.GetValue<string>();
+```
+
+### Cast operators
+No explicit or implicit cast operators exist since `GetValue<T>` allows for a single, consistent programming model. This also works on not just the "known" types in the CLR but also custom data types known by the serializer.
+
+Cast operators do not work in all languages such as F# and even for C# would not work nicely since `System.Object` is allowed for `JsonArray` and `JsonCollection` elements. For example:
+```cs
+JsonArray jArray = ...
+jArray[2] = "hello";
+// Is jArray[2] a string or a JsonValue?
+
+jArray[2] = 2;
+// jArray[2] is always an int since we can't have implicit cast operators for number types.
+```
+
+Although it is possible to add cast operators for `JsonValue` to\from string\bool, number types should not have any implicit operator since they may throw - for example, an internal value of "3.14" cannot be converted into `int` so an exception must be thrown.
```

Comment:
In the scenario you've mentioned why would 3.14 need to be converted to int?

Reply:
STJ supports quoted numbers meaning numbers can be stored as string in JSON but at run-time the floating\integral values are desired.